### ISSUE TITLE
undefined Method "save_granular"

### ISSUE THREAD

**Issue by DanielDocekalFHTW** [Shopify/graphql_java_gen#42]
Basically i copied the code from the README

```require 'graphql_java_gen'
require 'graphql_schema'
require 'json'

introspection_result = File.read("graphql_schema.json")
schema = GraphQLSchema.new(JSON.parse(introspection_result))

GraphQLJavaGen.new(schema,
  package_name: "com.example.MyApp",
  nest_under: 'ExampleSchema',
  custom_scalars: [
    GraphQLJavaGen::Scalar.new(
      type_name: 'Decimal',
      java_type: 'BigDecimal',
      deserialize_expr: ->(expr) { "new BigDecimal(jsonAsString(#{expr}, key))" },
      imports: ['java.math.BigDecimal'],
    ),
  ]
).save_granular("#{Dir.pwd}/../MyApp/src/main/java/com/example/MyApp/")

```

I changed the filenames and paths, i installed the gems. If i run the file though commandline says 

javagen.rb:19:in `<main>': undefined method `save_granular' for #<GraphQLJavaGen:0x0000000003234710> (NoMethodError)

Maybe it is my mistake but i never had anything to do with ruby and have no clue.

Thank you for your help.

**Comment by dylanahsmith** [comment]
That example is for the latest master, not for the last release.  I think that is why you are having problems

**Comment by DanielDocekalFHTW** [comment]
@dylanahsmith do you know when the next release is planed?
If it is not in the forseeable future, is there an documentation on how to build the project so that this example works ?

**Comment by dylanahsmith** [comment]
You could look at the example in the released version: https://github.com/Shopify/graphql_java_gen/tree/v0.1.1

If you want to use the latest master, then you would probably want to use a git submodule so that you keep the code generator and support code in sync.  If you are doing that, then you can reference the ruby gem by path as mentioned in the [development section of the README](https://github.com/Shopify/graphql_java_gen#development)

I haven't been involved in mobile development lately, so I'm not sure when the next planned release would be.

**Comment by DanielDocekalFHTW** [comment]
So i downloaded the master branch

Started a new Eclipse project where i inserted a .rb file with the code from the readme into the root

At the root of the project there is also schema.json and the gemfile with 
```
source 'https://rubygems.org'

gem 'graphql_java_gen' , path: 'C:\Users\MyName\Downloads\graphql_java_gen-master'
```
also an build.gradle

I have build the gradle file, i ran the command bundle inside the root and i can successfully run the code that throws everything into one file, but i am still unable to run the Code for saving into mulitple files without the Error mentioned in the OP.

Did i miss any steps ?

Thank you for any advice :)

**Comment by dylanahsmith** [comment]
If you got the same error, then it still sounds like you still aren't using the latest version.

Did you run ruby with `bundle exec ruby` so that the dependency in the Gemfile is getting used?  Alternatively, you could use `require "bundler/setup"` in the script before all the other `require` statements to avoid the need to do that.

Also, you should be able to use a relative path in the Gemfile so that you don't have to hard code your user specific path.  For more information on writing the Gemfile or using bundler, the documentation for it might be helpful: https://bundler.io/

**Comment by DanielDocekalFHTW** [comment]
executed bundle exec ruby  in my root it took for ever and did nothing until i interrupted,
the require "bundler/setup" finally worked and got everything in an extra file.

Thank you very much for sticking around @dylanahsmith

**Comment by DanielDocekalFHTW** [comment]
@dylanahsmith  do you or anybody else know why i get the error that the class(es) don't know where to import
```
com.shopify.graphql.support.*
com.google.gson.
```
from?
Exact error is that those libraries cant be resolved

my build.gradle file looks like 
```
apply plugin: 'java-library'
apply plugin: 'java'
apply plugin: 'eclipse'

repositories {
    google()
    jcenter()
}

dependencies {

    api 'org.apache.commons:commons-math3:3.6.1'
    implementation 'com.google.guava:guava:23.0'
    testImplementation 'junit:junit:4.12'
    implementation 'com.shopify.graphql.support:support:0.2.0'
    implementation 'com.google.code.gson:gson:2.7'
}

```

**Comment by dylanahsmith** [comment]
As I said before

> If you want to use the latest master, then you would probably want to use a git submodule so that you keep the code generator and support code in sync.

so you need to build the java package in the support directory from source.

We created a `GraphQLSupport` symlink for this pointing to the actual directory `Vendor/graphql_java_gen/support`.

in settings.gradle for the project we added `':GraphQLSupport'` to the `include` line.

In build.gradle for the application we have

```
dependencies {
    // ...
    compile project(':GraphQLSupport')
```

**Comment by DanielDocekalFHTW** [comment]
TLDR; i worked around that and just took all the Java Classes from the git i downloaded and inserted it into the Project manually.

now at the moment i'm trying to test the generated Class(es) but i m unable to add Input arguments, and i cant find any documentation on that topic.
´´´
String queryString = ExampleSchema.query(query -> query
				       .allOfferings(?, offering->offering.id())
				   ).toString();
			System.out.println(queryString); 
´´´

If you could tell me what to insert at the "?"(second Line) i tried quiet a lot of methods and interfaces from the generated code but none work.

**Comment by dylanahsmith** [comment]
Its hard to tell you exactly what you need, since that depends on your GraphQL schema.  E.g. if the argument is an input object, then you would construct an instance of the class generated for the input object.

**Comment by DanielDocekalFHTW** [comment]
Well it says there should be filled in an AllOfferingArgumentsDefinition, this is an Interface so i cant create an instance even if it only has one method define()

´´´ 
public interface AllOfferingsArgumentsDefinition {
            void define(AllOfferingsArguments args);
}
´´´

if i try creating  AllOfferingArgumentsDefinition it says i cant use it in a static way
´´´
 public class AllOfferingsArguments extends Arguments {
            public AllOfferingsArguments(StringBuilder _queryBuilder) {
                super(_queryBuilder, true);
            }

            public AllOfferingsArguments offeringCategoryUri(String value) {
                if (value != null) {
                    startArgument("offeringCategoryUri");
                    Query.appendQuotedString(_queryBuilder, value.toString());
                }
                return this;
            }

            public AllOfferingsArguments onlyActive(Boolean value) {
                if (value != null) {
                    startArgument("onlyActive");
                    _queryBuilder.append(value);
                }
                return this;
            }
}
´´´

here is the Method in question
´´´
public ExchangeQueriesQuery allOfferings(AllOfferingsArgumentsDefinition argsDef, OfferingQueryDefinition queryDef) {
            startField("allOfferings");

            AllOfferingsArguments args = new AllOfferingsArguments(_queryBuilder);
            argsDef.define(args);
            AllOfferingsArguments.end(args);

            _queryBuilder.append('{');
            queryDef.define(new OfferingQuery(_queryBuilder));
            _queryBuilder.append('}');

            return this;
}
´´´

**Comment by dylanahsmith** [comment]
Oh, that is for optional arguments.  You can use a lambda expression to specify those, allowing you to only specify the arguments you want specified in the query.  E.g. `args -> args. offeringCategoryUri("http://example.com/").onlyActive(false)`
